### PR TITLE
Virt handler fixes and cleanup

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
The virt-handler suffers of multiple possible race conditions during shutdown:
1. There isn't a correct sync between the controller and the heartbeat shutdown
2. There isn't proper error handling of the final node labelling process.

#### For the first one:
The intent of the `heartBeatDone` channel was
to ensure the heartbeat is finished
before stopping the handler vm controller.
In particular, its intention was to prevent
a shutdown race:
- heartbeat tries to stop and do a last http request
- controller shuts down
If the controller shuts down first, then the main function
just exists, killing all remaining go routines right away.

Actually it is not doing its job since
`c.heartBeat.Run(c.heartBeatInterval, stopCh)` is
not a blocking function (it spawns a goroutine and
immediately returns the channel).
As result, the `heartBeat` is always immediately
closed after the start of the heartBeat goroutine.

To ensure that the heartbeat is completed before shutting
down the controller, we should block on the channel returned
by the heartbeat itself.

#### For the second one:
When the heartbeat is notified of a shutdown
by the stopchan, it marks the node unschedulable
before returning and thus end the entire process.

This labelling process is performed in a goroutine
which at the end signal the job done closing the
`done` channel.

This logic actually is not necessary, since it can
be a simple body function, without the goroutine.

In addition, there isn't a proper error handling.
In an error occurs during the goroutine execution
(i.e. node patch failure), nothing is returned
and the channel is never closed, which cause the
`<-labelNodeDone` to hangs and leave the virt-handler
stuck.

With this commit both problem are handler:
1. drop the unneeded goroutine
2. handle the error retrying it for a `DefaultBackoff`
(https://github.com/kubernetes/client-go/blob/37045084c2aa82927b0e5ffc752861430fd7e4ab/util/retry/util.go#L38)

Lastly, there is an optimization on the guest agent command support
calculation.

### Release note

```release-note
Bugfix: fix possible virt-handler race condition and stuck situation during shutdown
```

### CODE REVIEW

File: pkg/virt-handler/vm.go [hunk 1]
```diff
@@ -1712,14 +1701,14 @@ func (cond *multipleNonMigratableCondition) generateStorageLiveMigrationConditio
 	}
 }
 
-func (d *VirtualMachineController) calculateLiveStorageMigrationCondition(vmi *v1.VirtualMachineInstance) *v1.VirtualMachineInstanceCondition {
+func (c *VirtualMachineController) calculateLiveStorageMigrationCondition(vmi *v1.VirtualMachineInstance) *v1.VirtualMachineInstanceCondition {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/vm.go:1615]: Yes, :slightly_smiling_face: just `c` for controller

File: pkg/virt-handler/guest-agent/support.go [hunk 2]
```diff
@@ -53,28 +53,22 @@ var PasswordRelatedGuestAgentCommands = []string{
...
-		for _, foundCmd := range commands {
-			if cmd == foundCmd.Name {
-				if foundCmd.Enabled {
-					found = true
-				}
-				break
-			}
-		}
-		if found == false {
+		if enabled, exists := availableCmdsMap[cmd]; !exists || !enabled {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/guest-agent/support.go:58]: Well, we should return true if every required command is also available and enabled.
Currently, the `if` is looking for something that is not available or enabled and early exit as soon as finds a match.
Your suggestion will return true as soon as one (positive) match is found, which does not guarantee that all the
commands are available and enabled  :)

File: pkg/virt-handler/guest-agent/support.go
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/guest-agent/support.go:1]: I am not sure. Actually they are already tested in `vm_test`

File: pkg/virt-handler/guest-agent/support.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/guest-agent/support.go:58]: Ah :man_facepalming: yeah I missed that. You're right, thanks!

File: pkg/virt-handler/vm.go [hunk 3]
```diff
@@ -1677,11 +1677,7 @@ func (c *VirtualMachineController) Run(threadiness int, stopCh chan struct{}) {
 		}
 	}
 
-	heartBeatDone := make(chan struct{})
-	go func() {
-		c.heartBeat.Run(c.heartBeatInterval, stopCh)
-		close(heartBeatDone)
-	}()
+	heartBeatDone := c.heartBeat.Run(c.heartBeatInterval, stopCh)
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/vm.go:1680]: I think we need it to guarantee that the `h.labelNodeUnschedulable()` is executed when the `stop` channel is closed.
The VM controller runs the heartBeat through a goroutine and continues its execution.

**-- Without the `heartBeatDone` --**
When the stop channel is closed, both heartBeat goroutine and VM controller continue their execution in parallel:
- the controller tries to end the whole function
- the heartBeat marks the node unschedulable and ends.

This is a race condition since there are two things that could happen:

- the heartBeat is fast enough and ends before the controller ends
   - node is properly marked unschedulable
  
- the controller ends before the heartBeat finishes its job and gets killed since the parent process is dead.
   - the heartBeat may or may have not marked the node unschedulable
   
**-- With the `heartBeatDone` --**
When the stop channel is closed, the VM controller waits for the end of the heartBeat before continue its execution and shutdown.

[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/vm.go:1680]: Does it make sense?

**Comment by fossedihelm** [comment]
[Addressed comments](https://github.com/kubevirt/kubevirt/compare/2f535214199b5c9d5ae99ae0a1009e9d07e2897a..8d3a14dbf4ad86e39261c96955ff050ce5a80f72) and [Rebase](https://github.com/kubevirt/kubevirt/compare/8d3a14dbf4ad86e39261c96955ff050ce5a80f72..06eae13c47dab8f1c90b8385bfe3b60a3f1ad3c2)
@iholder101 PTAL, thanks

**Review by jean-edouard — COMMENTED** [review]
/approve
Very nice, thanks!
Just one question, why `guest-agent/support.go` and not just `guest-agent.go`? Do we plan on adding more files to that directory?
Also if the guest agent functions were in the same package as the rest of virt-handler, we wouldn't need to prefix all calls with `guest_agent.` (is it right to use an underscore there btw? Looks like an unusual package name)

**Comment by fossedihelm** [comment]
> /approve Very nice, thanks! Just one question, why `guest-agent/support.go` and not just `guest-agent.go`? Do we plan on adding more files to that directory? Also if the guest agent functions were in the same package as the rest of virt-handler, we wouldn't need to prefix all calls with `guest_agent.` (is it right to use an underscore there btw? Looks like an unusual package name)

@jean-edouard Right! I will drop the directory in favor of a simple file :+1: Thank you!

File: pkg/virt-handler/vm.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-handler/vm.go:1694]: Yes, it does :) thank you!

File: pkg/virt-handler/vm.go [hunk 4]
```diff
@@ -1704,8 +1700,8 @@ func (c *VirtualMachineController) Run(threadiness int, stopCh chan struct{}) {
 		go wait.Until(c.runWorker, time.Second, stopCh)
 	}
 
-	<-stopCh
 	<-heartBeatDone
+	<-stopCh
```
Comment (iholder101) [inline pkg/virt-handler/vm.go:1704]: nit: why is this needed? I'm pretty sure that the behavior is equivalent to before your changes.

**Review by iholder101 — COMMENTED** [review]
Thank you @fossedihelm!
/lgtm

File: pkg/virt-handler/vm.go (hunk 4 above)
Comment (fossedihelm) [reply pkg/virt-handler/vm.go:1704]: Yes, the behavior is equivalent.
I do that to highlight that the heartbeat should succeed before moving on. :)
Thank you!
I will rollback if you think it's misleading

**Comment by brianmcarey** [comment]
/override pull-kubevirt-e2e-k8s-1.31-sig-performance

Triggered by status-reconciler when lane was bumped.
